contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,345 @@
+)
+```
+
+PreStopHandler is introduced to aggregate Handler and TerminationReasonDelivery.
+
+```golang
+// PreStopHandler aggregates Handler and a TerminationReasonDelivery to allow 
+// for configuration of the delivery method for a termination reason consumed 
+// by the hook.
+type PreStopHandler struct {
+    Handler
+    // ReasonDelivery provides configuration for the delivery of a termination 
+    // reason to the PreStopHandler. If nil, the termination reason will be 
+    // delivered to the preStop lifecycle hook by setting the
+    // KUBE_POD_TERM_REASON environment variable to the value of the termination
+    // reason.
+    ReasonDelivery *TerminationReasonDelivery `json:"reasonDelivery,ommitempty"`
+}
+```
+
+The Lifecycle struct is modified such that `PreStop` is a PreStopHandler.
+
+```golang
+type Lifecycle struct {
+    // Other fields and existing comments omitted for brevity.
+ 
+    PreStop *PreStopHandler
+}
+```
+
+## API Server 
+During its graceful delete processing, When a termination reason is specified 
+via DeleteOptions, the API server will include the reason in the 
+Pod's metadata when setting the Pod's `DeletionTimestamp`.
+
+### Validation
+In addition to the existing validation performed for the Lifecycle struct, the 
+API Server should fail validation if `PreStop` is not nil, and if one of the 
+following are true.
+
+1. `PreStop` indicates a command action and `.ReasonDelivery.Header` is not nil.
+1. `PreStop` indicates a HTTP GET action and `.ReasonDelivery.Env` is not nil.
+1. Both `.ReasonDelivery.Header` and `.ReasonDelivery.Env` are not nil.
+1. `.ReasonDelivery.Env` is not nil and it points to string that is not a 
+valid environment variable.
+
+### Pod Deletion
+When the API Server performs its graceful delete processing, in addition to 
+setting the `DeletionTimestamp` of the subject ObjectMeta, if the DeleteOptions 
+contains a `Reason`, the `Reason` should be copied to the `TerminationReason` 
+field of the ObjectMeta.
+
+## Kubelet
+Kubelet will supply a termination reason via a configured pre-stop lifecycle 
+hook under the following conditions.
+
+1. The Pod has been explicitly deleted with a supplied `TerminationReason`.
+1. The Pod is the target of an eviction.
+1. The Pod is being terminated due to an intolerable taint.
+
+In all cases Kubelet will deliver the termination reason to a container 
+with a declared `PreStop` handler in accordance with the configuration of its 
+`ReasonDelivery`.
+
+### Termination Reason Delivery
+
+When Kublet processes the `PreStop` Handlers of a Pod's containers, prior to 
+sending a `TERM` signal and starting the termination grace timer, if a 
+termination reason is to be delivered, Kubelet will do the following.
+
+1. If the `PreStop` handler indicates a command action, Kubelet will supply the 
+termination reason to the container based on the following criteria.
+   1. If the `ReasonDelivery` is nil Kubelet will set the 
```

Comment:
Does it make sense to put the logic for handling default behaviors from Kubelet to the API (default.go) to keep the Kubelet a little bit simpler?